### Making exon methylation level graphs ###


setwd("~/Dropbox/PhD Folder/Virtual Lab Book/MERN_Analysis/Meth_vs_Exp/methylation_over_exons")

library(ggplot2)
library(ggsignif)
library(ggthemes)
library(plyr)
library(reshape2)
library(car)


# Read in the exon information file, this was created by the script: mern_meth_gff_annotation.R
# grep was then used to take out only rows containing exon as a feature and only CpG positions (less than 200 non-CpG were removed)
# this file contains methylation values for EVERY CpG in the genome that overlaps an exon 
exon_info <- read_delim("~/Dropbox/PhD Folder/Virtual Lab Book/MERN_Analysis/Meth_vs_Exp/methylation_over_exons/exon_information_CpG_only.csv", 
                                               "\t", escape_double = FALSE, col_names = FALSE, 
                                               trim_ws = TRUE)

# Cut down the file to useable information
exon_info<-exon_info[,-c(1,2,3,5,6,7,10)]
colnames(exon_info)<-c("meth_proportion","geneID","exon_number","status")

# Make a new column contianing geneID and exon_number as a unique identifier per exon
exon_info$exon_id<-paste(exon_info$geneID, "_",exon_info$exon_number)
exon_info$exon_id <- gsub('\\s+', '', exon_info$exon_id)
exon_info<-exon_info[,-c(2,3)]

# Take the mean methylation level of every exon
mean_exon_meth<-aggregate( meth_proportion ~ exon_id+status, exon_info, mean )
rm(exon_info)


# Read in the list of genes containing diff exon expression generated by DEXseq, 86 total exons diff expressed
diff_exp_exons<- read_csv("~/Dropbox/PhD Folder/Virtual Lab Book/MERN_Analysis/Meth_vs_Exp/methylation_over_exons/significant_exon_useage_withoutj824.csv")

# Cut down to just useable information
diff_exp_exons<-diff_exp_exons[,c(1,2,10)]

# Make columns compatible with previous data frame
diff_exp_exons$featureID <- gsub('E0', '', diff_exp_exons$featureID)
diff_exp_exons$featureID<-as.numeric(diff_exp_exons$featureID)

colnames(diff_exp_exons)<-c("geneID","exon_number","Log2FC_repro_to_nonrepro")

diff_exp_exons$exon_id<-paste(diff_exp_exons$geneID, "_",diff_exp_exons$exon_number)
diff_exp_exons$exon_id <- gsub('\\s+', '', diff_exp_exons$exon_id)
diff_exp_exons<-diff_exp_exons[,-c(1,2)]


# Merge everything to get an exon list with methylation level and differential expression status
DEE_with_meth<-merge(diff_exp_exons, mean_exon_meth, by="exon_id")
DEE_with_meth$Log2FC_repro_to_nonrepro<-as.numeric(DEE_with_meth$Log2FC_repro_to_nonrepro)
DEE_with_meth$meth_proportion<-as.numeric(DEE_with_meth$meth_proportion)
DEE_with_meth$status<-as.factor(DEE_with_meth$status)

# Make new column so plotting is easier
DEE_with_meth$exp_and_status[DEE_with_meth$Log2FC_repro_to_nonrepro<0 & DEE_with_meth$status=="repro" ]<-"skipped"
DEE_with_meth$exp_and_status[DEE_with_meth$Log2FC_repro_to_nonrepro>0 & DEE_with_meth$status=="non_repro"]<-"skipped"
DEE_with_meth$exp_and_status[DEE_with_meth$Log2FC_repro_to_nonrepro>0 & DEE_with_meth$status=="repro" ]<-"included"
DEE_with_meth$exp_and_status[DEE_with_meth$Log2FC_repro_to_nonrepro<0 & DEE_with_meth$status=="non_repro"]<-"included"

# Make proportion into percentage for comparison with other graphs
DEE_with_meth$meth_proportion<-DEE_with_meth$meth_proportion*100




## Dot plot
dot_data<-subset(DEE_with_meth, !status=='overall_mean')
dot_data_skipped<-subset(dot_data, exp_and_status=="skipped")
skipped_repro<-subset(dot_data_skipped, status=="repro")
skipped_nonrepro<-subset(dot_data_skipped, status=="non_repro")
dot_data_included<-subset(dot_data, exp_and_status=="included")
included_repro<-subset(dot_data_included, status=="repro")
included_nonrepro<-subset(dot_data_included, status=="non_repro")

skipped_repro$name<-"skipped_repro"
skipped_repro<-skipped_repro[,c(4,6)]
skipped_nonrepro$name<-"skipped_nonrepro"
skipped_nonrepro<-skipped_nonrepro[,c(4,6)]
included_repro$name<-"included_repro"
included_repro<-included_repro[,c(4,6)]
included_nonrepro$name<-"included_nonrepro"
included_nonrepro<-included_nonrepro[,c(4,6)]

dot_data1<-rbind(skipped_repro,skipped_nonrepro,included_repro,included_nonrepro)
dot_data1$status<-dot_data1$name
dot_data1$status<-gsub("skipped_","", dot_data1$status)
dot_data1$status<-gsub("included_","", dot_data1$status)
dot_data1$log_meth<-log(dot_data1$meth_proportion)

condifence_intervals <- function(x) {
  m <- mean(x)
  sd1 <- sd(x)
  n <- length(x)
  error <- qnorm(0.95)*sd1/sqrt(n)
  ymin <- m-error
  ymax <- m+error
  return(c(y=m,ymin=ymin,ymax=ymax))
}

ggplot(dot_data1, aes(x=name, y=log_meth, fill=status)) + 
  geom_violin(trim = FALSE)+
  geom_dotplot(binaxis='y', stackdir='center', binwidth = 0.2, dotsize = 0.2, fill="black")+
  stat_summary(fun.data=condifence_intervals, color="blue",size=1)+
  xlab("Gene Group")+
  ylab("log(Average Percentage CpGs Methylated)")+
  scale_fill_manual("", breaks=c("repro", "nonrepro"),
                    values = c("grey","grey35"),
                    labels= c("Repro", "Sterile"))+
  scale_x_discrete(limits=c("included_nonrepro", "included_repro","skipped_nonrepro", "skipped_repro"),
                   labels=c("Included", "Included", "Skipped", "Skipped"))+
  theme_bw()+
  theme(axis.text=element_text(size=26),
        axis.title=element_text(size=28),
        legend.text=element_text(size=28))

#ggplot(dot_data1, aes(x=name, y=methylation)) + 
#  geom_violin(trim = FALSE)+
#  geom_dotplot(binaxis='y', stackdir='center', binwidth = 0.5, dotsize = 0.5)+
#  stat_summary(fun.data=condifence_intervals, color="blue")

#ggplot(dot_data1, aes(x=name, y=log_meth)) + 
#  geom_dotplot(binaxis='y', stackdir='center', binwidth = 0.03, dotsize = 0.3)+
#  stat_summary(fun.data=mean_sdl, fun.args = list(mult=2), 
#               geom="pointrange", color="red")

# Do a t-test on the two groups
t.test(DEE_with_meth$meth_proportion~DEE_with_meth$exp_and_status) 

#Welch Two Sample t-test

#data:  DEE_with_meth$meth_proportion by DEE_with_meth$exp_and_status
#t = 0.10862, df = 139.47, p-value = 0.9137
#alternative hypothesis: true difference in means is not equal to 0
#95 percent confidence interval:
#  -6.229843  6.954183
#sample estimates:
#  mean in group included  mean in group skipped 
#7.620994               7.258824 


## ---------------------------------------------------
# Pretty sure I shouldn't be doing a t-test as doesn't take repro status into account
## ---------------------------------------------------

# Make a variable which is DEG/nonDEG
head(dot_data1)
dot_data1$name <- gsub("included.*","included", dot_data1$name)
dot_data1$name <- gsub("skipped.*","skipped", dot_data1$name)

# ...looking online a factorial anova?
# (http://rcompanion.org/handbook/G_09.html)

interaction.plot(x.factor     = dot_data1$name,
                 trace.factor = dot_data1$status, 
                 response     = dot_data1$meth_proportion, 
                 fun = mean,
                 type="b",
                 col=c("black","red"),  
                 pch=c(19, 17),
                 leg.bty = "o")

model = lm(methylation ~ name*status,
           data = dot_data1)
Anova(model,
      type = "II") # there is no interaction effect so type II appropriate?


# If the above is the right thing to do then following Eamonn's previous code ...


model1<-lm(meth_proportion~name*status, data=dot_data1)
model2<-lm(meth_proportion~name+status, data=dot_data1)

anova(model1,model2) # This tell us the interaction isn't significant so proceed to type II anova

#Analysis of Variance Table

#Model 1: meth_proportion ~ name * status
#Model 2: meth_proportion ~ name + status
#Res.Df   RSS Df Sum of Sq      F Pr(>F)
#1    138 55221                           
#2    139 55238 -1   -17.101 0.0427 0.8365

summary.lm(model2) # Get the same result as above code

#Call:
#  lm(formula = meth_proportion ~ name + status, data = dot_data1)

#Residuals:
#  Min     1Q Median     3Q    Max 
#-7.977 -7.276 -7.016 -6.341 88.052 

#Coefficients:
#  Estimate Std. Error t value Pr(>|t|)   
#(Intercept)   7.4024     2.7045   2.737  0.00701 **
#  nameskipped  -0.4998     3.4460  -0.145  0.88489   
#statusrepro   0.5749     3.4460   0.167  0.86774   
#---
#  Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

#Residual standard error: 19.93 on 139 degrees of freedom
#Multiple R-squared:  0.0002845,	Adjusted R-squared:  -0.0141 
#F-statistic: 0.01978 on 2 and 139 DF,  p-value: 0.9804
drop1(model2, test="F")

#Single term deletions

#Model:
#  meth_proportion ~ name + status
#Df Sum of Sq   RSS    AIC F value Pr(>F)
#<none>              55238 852.83               
#name    1    8.3604 55247 850.85  0.0210 0.8849
#status  1   11.0610 55249 850.86  0.0278 0.8677
