### Making exon methylation level graphs ###


setwd("~/Dropbox/PhD Folder/Virtual Lab Book/MERN_Analysis/Meth_vs_Exp/methylation_over_exons")

library(ggplot2)
library(ggsignif)
library(ggthemes)
library(plyr)
library(reshape2)
library(car)


# Read in the exon information file, this was created by the script: mern_meth_gff_annotation.R
# grep was then used to take out only rows containing exon as a feature and only CpG positions (less than 200 non-CpG were removed)
# this file contains methylation values for EVERY CpG in the genome that overlaps an exon 
exon_info <- read_delim("~/Dropbox/PhD Folder/Virtual Lab Book/MERN_Analysis/Meth_vs_Exp/methylation_over_exons/exon_information_CpG_only.csv", 
                        "\t", escape_double = FALSE, col_names = FALSE, 
                        trim_ws = TRUE)

# Cut down the file to useable information
exon_info<-exon_info[,-c(1,2,3,5,6,7,10)]
colnames(exon_info)<-c("meth_proportion","geneID","exon_number","status")

# Make a new column contianing geneID and exon_number as a unique identifier per exon
exon_info$exon_id<-paste(exon_info$geneID, "_",exon_info$exon_number)
exon_info$exon_id <- gsub('\\s+', '', exon_info$exon_id)

# Take the mean methylation level of every exon
mean_exon_meth<-aggregate( meth_proportion ~ exon_id+status+exon_number+geneID, exon_info, mean )
rm(exon_info)


# Read in the list of genes containing diff exon expression generated by DEXseq, 86 total exons diff expressed
diff_exp_exons<- read_csv("~/Dropbox/PhD Folder/Virtual Lab Book/MERN_Analysis/Meth_vs_Exp/methylation_over_exons/significant_exon_useage_withoutj824.csv")

# Cut down to just useable information
diff_exp_exons<-diff_exp_exons[,c(1,2,10)]

# Make columns compatible with previous data frame
diff_exp_exons$featureID <- gsub('E0', '', diff_exp_exons$featureID)
diff_exp_exons$featureID<-as.numeric(diff_exp_exons$featureID)

colnames(diff_exp_exons)<-c("geneID","exon_number","Log2FC_repro_to_nonrepro")

diff_exp_exons$exon_id<-paste(diff_exp_exons$geneID, "_",diff_exp_exons$exon_number)
diff_exp_exons$exon_id <- gsub('\\s+', '', diff_exp_exons$exon_id)


# Make a list of exons that are not differentially expressed 
all_exons<-as.data.frame(mean_exon_meth[,1])
colnames(all_exons)<-c("exon_id")
non_diff_exp_exons<-as.data.frame(all_exons[!(all_exons$exon_id %in% diff_exp_exons$exon_id),1])
colnames(non_diff_exp_exons)<-c("exon_id")
rm(all_exons)


# Merge everything to get an exon list with methylation level and differential expression status
DEE_with_meth<-merge(diff_exp_exons, mean_exon_meth, by="exon_id")
DEE_with_meth$name<-"DEE"
DEE_with_meth<-DEE_with_meth[,-c(6,7)]
colnames(DEE_with_meth)<-c("exon_id","geneID","exon_number","Log2FC_repro_to_nonrepro","status","meth_proportion","name")

non_DEE_with_meth<-merge(non_diff_exp_exons, mean_exon_meth, by="exon_id")
non_DEE_with_meth<-non_DEE_with_meth[!is.na(non_DEE_with_meth$meth_proportion),]
non_DEE_with_meth$name<-"non_DEE"

# Take a quick look at the mean methylation of these exons
non_DEE_with_meth_mean<-mean(non_DEE_with_meth$meth_proportion) # 0.12
DEE_with_meth_mean<-sapply(split(DEE_with_meth$meth_proportion, DEE_with_meth$status), mean)# similar 0.072 0.076

# Put it all together for plotting
non_DEE_with_meth$Log2FC_repro_to_nonrepro<-"N/A"
final_data<-rbind(non_DEE_with_meth,DEE_with_meth)
final_data$status<-gsub("non_repro","nonrepro",final_data$status)

# Make proportions into percentages to keep all graphs the same
final_data$meth_proportion<-final_data$meth_proportion*100

## Dot plot
dot_data<-subset(final_data, !status=='overall_mean')
dot_data_DEE<-subset(dot_data, name=="DEE")
DEE_repro<-subset(dot_data_DEE, status=="repro")
DEE_nonrepro<-subset(dot_data_DEE, status=="nonrepro")
dot_data_nonDEE<-subset(dot_data, name=="non_DEE")
nonDEE_repro<-subset(dot_data_nonDEE, status=="repro")
nonDEE_nonrepro<-subset(dot_data_nonDEE, status=="nonrepro")

DEE_repro$name<-"DEE_repro"
DEE_repro<-DEE_repro[,c(5,6)]
DEE_nonrepro$name<-"DEE_nonrepro"
DEE_nonrepro<-DEE_nonrepro[,c(5,6)]
nonDEE_repro$name<-"nonDEE_repro"
nonDEE_repro<-nonDEE_repro[,c(5,6)]
nonDEE_nonrepro$name<-"nonDEE_nonrepro"
nonDEE_nonrepro<-nonDEE_nonrepro[,c(5,6)]

dot_data1<-rbind(DEE_repro,DEE_nonrepro,nonDEE_repro,nonDEE_nonrepro)
dot_data1$status<-dot_data1$name
dot_data1$status<-gsub("nonDEE_","", dot_data1$status)
dot_data1$status<-gsub("DEE_","", dot_data1$status)
dot_data1$log_meth<-log(dot_data1$meth_proportion)

condifence_intervals <- function(x) {
  m <- mean(x)
  sd1 <- sd(x)
  n <- length(x)
  error <- qnorm(0.95)*sd1/sqrt(n)
  ymin <- m-error
  ymax <- m+error
  return(c(y=m,ymin=ymin,ymax=ymax))
}

ggplot(dot_data1, aes(x=name, y=log_meth, fill=status)) + 
  geom_violin(trim = FALSE)+
  geom_dotplot(binaxis='y', stackdir='center', binwidth = 0.007, dotsize = 0.3)+
  stat_summary(fun.data=condifence_intervals, color="blue",size=1)+
  xlab("Gene Group")+
  ylab("log(Average Percentage CpGs Methylated)")+
  scale_fill_manual("", breaks=c("repro", "nonrepro"),
                    values = c("grey","grey35"),
                    labels= c("Repro", "Sterile"))+
  scale_x_discrete(limits=c("DEE_nonrepro", "DEE_repro","nonDEE_nonrepro", "nonDEE_repro"),
                   labels=c("DEE", "DEE", "Non-DEE", "Non-DEE"))+
  theme_bw()+
  theme(axis.text=element_text(size=26),
        axis.title=element_text(size=28),
        legend.text=element_text(size=28))

#ggplot(dot_data1, aes(x=name, y=methylation)) + 
#  geom_violin(trim = FALSE)+
#  geom_dotplot(binaxis='y', stackdir='center', binwidth = 0.5, dotsize = 0.5)+
#  stat_summary(fun.data=condifence_intervals, color="blue")

#ggplot(dot_data1, aes(x=name, y=log_meth)) + 
#  geom_dotplot(binaxis='y', stackdir='center', binwidth = 0.03, dotsize = 0.3)+
#  stat_summary(fun.data=mean_sdl, fun.args = list(mult=2), 
#               geom="pointrange", color="red")

# Do a t-test on the two groups
t.test(final_data$meth_proportion~final_data$name) 

#Welch Two Sample t-test

#data:  final_data$meth_proportion by final_data$name
#t = -2.7747, df = 141.16, p-value = 0.006274
#alternative hypothesis: true difference in means is not equal to 0
#95 percent confidence interval:
#  -7.895698 -1.325650
#sample estimates:
#  mean in group DEE mean in group non_DEE 
#7.439909             12.050583 

## ---------------------------------------------------
# Pretty sure I shouldn't be doing a t-test as doesn't take repro status into account
## ---------------------------------------------------

# Make a variable which is DEG/nonDEG
tail(dot_data1)
dot_data1$name <- gsub("DEE.*","DEE", dot_data1$name)

# ...looking online a factorial anova?
# (http://rcompanion.org/handbook/G_09.html)

interaction.plot(x.factor     = dot_data1$name,
                 trace.factor = dot_data1$status, 
                 response     = dot_data1$meth_proportion, 
                 fun = mean,
                 type="b",
                 col=c("black","red"),  
                 pch=c(19, 17),
                 leg.bty = "o")

model = lm(methylation ~ name*status,
           data = dot_data1)
Anova(model,
      type = "II") # there is no interaction effect so type II appropriate?


# If the above is the right thing to do then following Eamonn's previous code ...


model1<-lm(meth_proportion~name*status, data=dot_data1)
model2<-lm(meth_proportion~name+status, data=dot_data1)

anova(model1,model2) # This tell us the interaction isn't significant so proceed to type II anova

#Analysis of Variance Table

#Model 1: meth_proportion ~ name * status
#Model 2: meth_proportion ~ name + status
#Res.Df       RSS Df Sum of Sq      F Pr(>F)
#1 343419 184650858                           
#2 343420 184650863 -1   -4.9567 0.0092 0.9235

summary.lm(model2) # Get the same result as above code

#Call:
#  lm(formula = meth_proportion ~ name + status, data = dot_data1)

#Residuals:
#  Min      1Q  Median      3Q     Max 
#-12.091 -11.766 -11.422  -2.913  88.055 

#Coefficients:
#  Estimate Std. Error t value Pr(>|t|)    
#(Intercept)  7.39908    1.94629   3.802 0.000144 ***
#  namenonDEE   4.61062    1.94629   2.369 0.017841 *  
#  statusrepro  0.08165    0.07914   1.032 0.302170    
#---
#  Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

#Residual standard error: 23.19 on 343420 degrees of freedom
#Multiple R-squared:  1.944e-05,	Adjusted R-squared:  1.362e-05 
#F-statistic: 3.338 on 2 and 343420 DF,  p-value: 0.0355

drop1(model2, test="F")

#Single term deletions

#Model:
#  meth_proportion ~ name + status
#Df Sum of Sq       RSS     AIC F value  Pr(>F)  
#<none>              184650863 2159195                  
#name    1   3017.36 184653880 2159199  5.6118 0.01784 *
#  status  1    572.42 184651435 2159194  1.0646 0.30217  
#---
#  Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
